src/hb-blob.cc

```diff
@@ -502,3 +506,113 @@ _try_writable (hb_blob_t *blob)
 
   return true;
 }
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+#include <windows.h>
+#include <io.h>
+
+#undef fstat
+#define fstat(a,b) _fstati64(a,b)
+#undef stat
+#define stat _stati64
+
+#ifndef S_ISREG
+# define S_ISREG(m) (((m) & _S_IFMT) == _S_IFREG)
+#endif
+#endif // defined(_WIN32) || defined(__CYGWIN__)
+
+#ifndef _O_BINARY
+# define _O_BINARY 0
+#endif
+
+#ifndef MAP_FAILED
+# define MAP_FAILED ((void *) -1)
+#endif
+
+struct hb_mapped_file_t
+{
+  char *contents;
+  unsigned long length;
+#if defined(_WIN32) || defined(__CYGWIN__)
+  HANDLE mapping;
+#endif
+};
+
+static void
+_hb_mapped_file_destroy (hb_mapped_file_t *file)
+{
+#ifdef HAVE_MMAP
+  munmap (file->contents, file->length);
+#elif defined(_WIN32) || defined(__CYGWIN__)
+  UnmapViewOfFile (file->contents);
+  CloseHandle (file->mapping);
+#else
+  free (file->contents);
+#endif
+
+  free (file);
+}
+
+/**
+ * hb_blob_create_from_file:
+ * @file_name: font filename.
+ *
+ * Returns: A hb_blob_t pointer with the content of the file
+ *
+ * Since: REPLACEME
+ **/
+hb_blob_t *
+hb_blob_create_from_file (const char *file_name)
+{
+  // Adopted from glib's gmappedfile.c with Matthias Clasen and
+  // Allison Lortie permission but changed to suit our need.
+  bool writable = false;
+  hb_memory_mode_t mm = HB_MEMORY_MODE_READONLY_MAY_MAKE_WRITABLE;
+
+  int fd = open (file_name, (writable ? O_RDWR : O_RDONLY) | _O_BINARY, 0);
+  if (fd == -1) return hb_blob_get_empty ();
+
+  hb_mapped_file_t *file = (hb_mapped_file_t *) malloc (sizeof (hb_mapped_file_t));
+
+  struct stat st;
+  if (fstat (fd, &st) == -1) goto fail;
+  // If the file size is 0, just give up, mmap doesn't like it also anyway
+  if (st.st_size == 0 && S_ISREG (st.st_mode)) goto fail;
+
+  file->length = (unsigned long) st.st_size;
+
+#ifdef HAVE_MMAP
+  file->contents = (char *) mmap (nullptr, file->length,
+				  writable ? PROT_READ|PROT_WRITE : PROT_READ,
+				  MAP_PRIVATE, fd, 0);
+  if (file->contents == MAP_FAILED) goto fail;
+#elif defined(_WIN32) || defined(__CYGWIN__)
+  file->mapping = CreateFileMapping ((HANDLE) _get_osfhandle (fd), nullptr,
+				     writable ? PAGE_WRITECOPY : PAGE_READONLY,
+				     0, 0, nullptr);
+  if (file->mapping == nullptr) goto fail;
+
+  file->contents = (char *) MapViewOfFile (file->mapping,
+					   writable ? FILE_MAP_COPY : FILE_MAP_READ,
+					   0, 0, 0);
+  if (file->contents == nullptr)
+  {
+    CloseHandle (file->mapping);
+    goto fail;
+  }
+#else
+  file->contents = (char *) malloc (file->length);
+  if (!file->contents) goto fail;
+  read (fd, file->contents, file->length);
+  mm = HB_MEMORY_MODE_WRITABLE;
+#endif
+
+  close (fd);
+  return hb_blob_create (file->contents, file->length, mm, (void *) file,
+			 (hb_destroy_func_t) _hb_mapped_file_destroy);
+
+fail:
```

Comment:
I think if memory-mapping fails we should still try reading the file. That would make it work for stdin, etc, no? Or there might be other reasons that mmap fails?

Reply:
In particular, accommodate for this piece of code in options.cc:
```

-      if (!len) {
-	/* GMappedFile is buggy, it doesn't fail if file isn't regular.
-	 * Try reading.
-	 * https://bugzilla.gnome.org/show_bug.cgi?id=659212 */
```

As long as running on a font file of `/dev/null` or `/dev/stdin` works, I'm fine. Thanks.

Reply:
> In particular, accommodate for this piece of code in options.cc:

I've picked the patch related to that comment https://bug659212.bugzilla-attachments.gnome.org/attachment.cgi?id=196855 

> As long as running on a font file of /dev/null or /dev/stdin works, I'm fine. Thanks.

They do, AFAICS

Reply:
Tested on Linux also.

Reply:
Thanks.